domains/client/domain-operator/src/fraud_proof.rs

```diff
@@ -262,61 +259,31 @@ where
         PCB: BlockT,
     {
         let block_hash = local_receipt.domain_block_hash;
-        let block_number = to_number_primitive(local_receipt.consensus_block_number);
-
+        let block_number = local_receipt.domain_block_number;
         let header = self.header(block_hash)?;
         let parent_header = self.header(*header.parent_hash())?;
 
-        // TODO: avoid the encode & decode?
-        let as_h256 = |state_root: &Block::Hash| {
-            H256::decode(&mut state_root.encode().as_slice())
-                .map_err(|_| FraudProofError::InvalidStateRootType)
-        };
-
         let prover = ExecutionProver::new(self.backend.clone(), self.code_executor.clone());
 
-        let parent_number = to_number_primitive(*parent_header.number());
-
-        let local_root = local_receipt
-            .execution_trace
-            .get(local_trace_index as usize)
-            .ok_or(FraudProofError::InvalidTraceIndex {
-                index: local_trace_index as usize,
-                max: local_receipt.execution_trace.len() - 1,
-            })?;
-
-        let consensus_parent_hash = H256::decode(
-            &mut self
-                .consensus_client
-                .header(local_receipt.consensus_block_hash)?
-                .ok_or_else(|| {
-                    sp_blockchain::Error::Backend(format!(
-                        "Header not found for consensus block {:?}",
-                        local_receipt.consensus_block_hash
-                    ))
-                })?
-                .parent_hash()
-                .encode()
-                .as_slice(),
-        )
-        .map_err(|_| FraudProofError::InvalidStateRootType)?;
-
-        let digest = Digest::default();
+        let digest = self.client.runtime_api().block_digest(block_hash)?;
```

Comment:
I'm sorry to confuse you on the digests. I have a question here. The digest you are fetching here is the post block digest. So it includes all the digest items that were added during the state transition and on_finalization. So the digests might not the same digests that were passed during the new block initialization then, correct ? So we may be adding the digest logs that will added in the later execution in the intialization itself. Thoughts @NingLin-P

Reply:
Okay, I took a closer look. I think the first version of this PR is fine, namely only including the `consensus_block_info` digest which is the only inherent digest and doesn't need the digest storage proof at all.

For `ApplyExtrinsic` and `FinalizeBlock`, the necessary extrinsic will be executed before generating the execution proof (see `prepare_storage_changes_before` and `prepare_storage_changes_before_finalize_block`) so the digests emitted during block execution/finalization will be included in the proof properly.

One last question about "there are digests from frontier we should consider here that are missing", do you mean the digest emitted during block execution/finalization or the inherent digest?